enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,161 @@
+
+## Motivation
+
+For consuming local storage from the nodes available in the OpenShift cluster a good amount of manual work has to be done for creating local PVs out of it, making hot-swapping of disks tough.
+
+### Goals
+
+- Automatically detection of available disks from nodes which can be used as PVs for OpenShift-cluster.
+- Should respond to the attach/detach events of the disks/devices.
+- Should have options for excluding a particular kind of disks.
+
+### Non-Goals
+
+## Proposal
+
+`local-storage-operator`is already capable of consuming local disks from the nodes and provisioning PVs out of it, but the disk/device paths needs to explicitly specified in the `LocalVolume` CR. The idea is to bring in one more layer of CR above it called `AutoDetectVolumes`.
+The controller for this CR will be responsible for automatically detecting disks/devices from the available nodes and creating/managing the life cycle of `LocalVolumes` which are governed by the `AutoDetectVolumes`.s
+
+### Risks and Mitigations
+
+- Detecting disks that contain data and/or are in-use.
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```
+type DeviceDiscoveryPolicyType string
+
+const (
+	raid DeviceDiscoveryPolicyType = "raid"
+	disk DeviceDiscoveryPolicyType = "disk"
+)
+
+type AutoDetectVolumeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              AutoDetectVolumeSpec   `json:"spec"`
+	Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+	// StorageClass name to use for set of matched devices
+	StorageClassName string `json:"storageClassName"`
+	// Volume mode. Raw or with file system
+	// + optional
+	VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+	// File system type
+	// +optional
+	FSType string `json:"fsType,omitempty"`
+	// Nodes on which the autoDetection must run
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// DeviceDiscoverPolicies are the list of policies that AutoDetectVolume's controller
+	// will be looking for during detection
+	DeviceDiscoveryPolicies []DeviceDiscoveryPolicy `json:"deviceDiscoveryPolicies,omitempty"`
+}
+
+type AutoDetectVolumeStatus struct {
+	Status string `json:"status"`
+}
+
+type DeviceDiscoveryPolicy struct {
+	// Device type that should be used for auto detection. This would be one of the types supported
```

Comment:
What about partitions?

Reply:
and logical volumes?

Reply:
If possible, I'd ask to just start with raw devices of whatever type we can easily detect. Just be careful to allow room for expanding functionality to support partitions and LVM artifacts.

Reply:
`lvm` and `part` are detected by the same effort. I think it would be effort to exclude them. That's based on my assumption that everything listed here are possible value's for `lsblk`'s TYPE column. 
If that's the case just add them to the doc's list @ashishranjan738 and mention that these are possible value's for `lsblk`'s TYPE column.

Reply:
Is the `DeviceType` optional or required field? Can you clarify?

Reply:
DeviceType is a required field.